src/DoctrineAuditBundle/DependencyInjection/Compiler/StorageConfigurationCompilerPass.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace DH\DoctrineAuditBundle\DependencyInjection\Compiler;
+
+use DH\DoctrineAuditBundle\AuditConfiguration;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+class StorageConfigurationCompilerPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->hasDefinition('dh_doctrine_audit.configuration')) {
+            return;
+        }
+
+        $config = $container->getParameter('dh_doctrine_audit.configuration');
+
+        if (isset($config['storage_entity_manager'])) {
+            $auditConfiguration = $container->get('dh_doctrine_audit.configuration');
+
+            if (null !== $auditConfiguration) {
+                $auditConfiguration = new AuditConfiguration(
+                    $config,
+                    $auditConfiguration->getUserProvider(),
+                    $auditConfiguration->getRequestStack(),
+                    $auditConfiguration->getFirewallMap(),
+                    $container->get($config['storage_entity_manager']),
+                    $auditConfiguration->getAnnotationLoader()
+                );
+
+                $container->set('dh_doctrine_audit.configuration', $auditConfiguration);
+            }
+        }
```

Comment:
What do you think of this ?

```
if (null === $config['storage_entity_manager']) {
    return;
}

$customEntityManager = $container->getDefinition($config['storage_entity_manager']);
if (!is_a($definition->getClass(), EntityManagerInterface::class, true)) {
    throw new InvalidArgumentException(sprintf('Service "%s" must be implement "%s".', $config['storage_entity_manager'], EntityManagerInterface::class));
}

$auditConfigurationDefinition = $container-> getDefinition('dh_doctrine_audit.configuration');
$auditConfigurationDefinition->replaceArgument(4, $customEntityManager);
// Or with a setter ? (safer)
// $auditConfigurationDefinition->addMethodCall('setEntityManager', $customEntityManager);
```

Reply:
Something's wrong in your snippet: `$definition->getClass()` => `$definition` is undefined.
But I like this check which makes sure `storage_entity_manager` is an entity manager.

Last, I don't want to use a setter to replace the entity manager in the `AuditConfiguration` service because even if its safer on one hand, it also lets anyone inject any entity manager object in the `AuditConfiguration` service at any time which can lead to unexpected behavior.

Reply:
You rigth, it's `$customEntityManager->getClass()`